// appium-xcode/build/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getVersion = exports.getPath = exports.getMaxTVOSSDKWithoutRetry = exports.getMaxTVOSSDK = exports.getMaxIOSSDKWithoutRetry = exports.getMaxIOSSDK = exports.getInstrumentsPath = exports.getConnectedDevices = exports.getCommandLineToolsVersion = exports.getClangVersion = exports.getAutomationTraceTemplatePathWithoutRetry = exports.getAutomationTraceTemplatePath = exports.default = exports.clearInternalCache = void 0;
require("source-map-support/register");
var xcode = _interopRequireWildcard(require("./lib/xcode"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
const {
  getPath,
  getVersion,
  getAutomationTraceTemplatePath,
  getMaxIOSSDK,
  getAutomationTraceTemplatePathWithoutRetry,
  getMaxIOSSDKWithoutRetry,
  getConnectedDevices,
  clearInternalCache,
  getInstrumentsPath,
  getCommandLineToolsVersion,
  getMaxTVOSSDK,
  getMaxTVOSSDKWithoutRetry,
  getClangVersion
} = xcode;
exports.getClangVersion = getClangVersion;
exports.getMaxTVOSSDKWithoutRetry = getMaxTVOSSDKWithoutRetry;
exports.getMaxTVOSSDK = getMaxTVOSSDK;
exports.getCommandLineToolsVersion = getCommandLineToolsVersion;
exports.getInstrumentsPath = getInstrumentsPath;
exports.clearInternalCache = clearInternalCache;
exports.getConnectedDevices = getConnectedDevices;
exports.getMaxIOSSDKWithoutRetry = getMaxIOSSDKWithoutRetry;
exports.getAutomationTraceTemplatePathWithoutRetry = getAutomationTraceTemplatePathWithoutRetry;
exports.getMaxIOSSDK = getMaxIOSSDK;
exports.getAutomationTraceTemplatePath = getAutomationTraceTemplatePath;
exports.getVersion = getVersion;
exports.getPath = getPath;
var _default = xcode;
exports.default = _default;
