// @appium/base-driver/build/lib/express/static.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.STATIC_DIR = void 0;
exports.guineaPig = guineaPig;
exports.guineaPigAppBanner = guineaPigAppBanner;
exports.guineaPigScrollable = guineaPigScrollable;
exports.welcome = welcome;

require("source-map-support/register");

var _path = _interopRequireDefault(require("path"));

var _logger = _interopRequireDefault(require("./logger"));

var _lodash = _interopRequireDefault(require("lodash"));

var _support = require("@appium/support");

var _bluebird = _interopRequireDefault(require("bluebird"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

let STATIC_DIR = _path.default.resolve(__dirname, '..', '..', '..', 'static');

exports.STATIC_DIR = STATIC_DIR;

if (_lodash.default.isNull(_path.default.resolve(__dirname).match(/build[/\\]lib[/\\]express$/))) {
  exports.STATIC_DIR = STATIC_DIR = _path.default.resolve(__dirname, '..', '..', 'static');
}

async function guineaPigTemplate(req, res, page) {
  const delay = parseInt(req.params.delay || req.query.delay || 0, 10);
  const throwError = req.params.throwError || req.query.throwError || '';
  let params = {
    throwError,
    serverTime: new Date(),
    userAgent: req.headers['user-agent'],
    comment: 'None'
  };

  if (req.method === 'POST') {
    params.comment = req.body.comments || params.comment;
  }

  _logger.default.debug(`Sending guinea pig response with params: ${JSON.stringify(params)}`);

  if (delay) {
    _logger.default.debug(`Waiting ${delay}ms before responding`);

    await _bluebird.default.delay(delay);
  }

  res.set('content-type', 'text/html');
  res.cookie('guineacookie1', 'i am a cookie value', {
    path: '/'
  });
  res.cookie('guineacookie2', 'cookié2', {
    path: '/'
  });
  res.cookie('guineacookie3', 'cant access this', {
    domain: '.blargimarg.com',
    path: '/'
  });
  res.send((await getTemplate(page))(params));
}

async function guineaPig(req, res) {
  return await guineaPigTemplate(req, res, 'guinea-pig.html');
}

async function guineaPigScrollable(req, res) {
  return await guineaPigTemplate(req, res, 'guinea-pig-scrollable.html');
}

async function guineaPigAppBanner(req, res) {
  return await guineaPigTemplate(req, res, 'guinea-pig-app-banner.html');
}

async function welcome(req, res) {
  let params = {
    message: "Let's browse!"
  };

  _logger.default.debug(`Sending welcome response with params: ${JSON.stringify(params)}`);

  res.send((await getTemplate('welcome.html'))(params));
}

async function getTemplate(templateName) {
  let content = await _support.fs.readFile(_path.default.resolve(STATIC_DIR, 'test', templateName));
  return _lodash.default.template(content.toString());
}
