// @appium/base-driver/build/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  BaseDriver: true,
  DriverCore: true,
  DeviceSettings: true,
  errorFromCode: true,
  DEFAULT_BASE_PATH: true,
  PROTOCOLS: true,
  STATIC_DIR: true,
  server: true,
  normalizeBasePath: true,
  JWProxy: true,
  getSummaryByCode: true,
  statusCodes: true,
  PREFIXED_APPIUM_OPTS_CAP: true,
  STANDARD_CAPS: true,
  processCapabilities: true,
  isStandardCap: true,
  validateCaps: true,
  DEFAULT_WS_PATHNAME_PREFIX: true
};
Object.defineProperty(exports, "BaseDriver", {
  enumerable: true,
  get: function () {
    return _driver.BaseDriver;
  }
});
Object.defineProperty(exports, "DEFAULT_BASE_PATH", {
  enumerable: true,
  get: function () {
    return _constants.DEFAULT_BASE_PATH;
  }
});
Object.defineProperty(exports, "DEFAULT_WS_PATHNAME_PREFIX", {
  enumerable: true,
  get: function () {
    return _websocket.DEFAULT_WS_PATHNAME_PREFIX;
  }
});
Object.defineProperty(exports, "DeviceSettings", {
  enumerable: true,
  get: function () {
    return _deviceSettings.DeviceSettings;
  }
});
Object.defineProperty(exports, "DriverCore", {
  enumerable: true,
  get: function () {
    return _core.DriverCore;
  }
});
Object.defineProperty(exports, "JWProxy", {
  enumerable: true,
  get: function () {
    return _proxy.JWProxy;
  }
});
Object.defineProperty(exports, "PREFIXED_APPIUM_OPTS_CAP", {
  enumerable: true,
  get: function () {
    return _capabilities.PREFIXED_APPIUM_OPTS_CAP;
  }
});
Object.defineProperty(exports, "PROTOCOLS", {
  enumerable: true,
  get: function () {
    return _constants.PROTOCOLS;
  }
});
Object.defineProperty(exports, "STANDARD_CAPS", {
  enumerable: true,
  get: function () {
    return _capabilities.STANDARD_CAPS;
  }
});
Object.defineProperty(exports, "STATIC_DIR", {
  enumerable: true,
  get: function () {
    return _static.STATIC_DIR;
  }
});
exports.default = void 0;
Object.defineProperty(exports, "errorFromCode", {
  enumerable: true,
  get: function () {
    return _protocol.errorFromMJSONWPStatusCode;
  }
});
Object.defineProperty(exports, "getSummaryByCode", {
  enumerable: true,
  get: function () {
    return _status.getSummaryByCode;
  }
});
Object.defineProperty(exports, "isStandardCap", {
  enumerable: true,
  get: function () {
    return _capabilities.isStandardCap;
  }
});
Object.defineProperty(exports, "normalizeBasePath", {
  enumerable: true,
  get: function () {
    return _server.normalizeBasePath;
  }
});
Object.defineProperty(exports, "processCapabilities", {
  enumerable: true,
  get: function () {
    return _capabilities.processCapabilities;
  }
});
Object.defineProperty(exports, "server", {
  enumerable: true,
  get: function () {
    return _server.server;
  }
});
Object.defineProperty(exports, "statusCodes", {
  enumerable: true,
  get: function () {
    return _status.codes;
  }
});
Object.defineProperty(exports, "validateCaps", {
  enumerable: true,
  get: function () {
    return _capabilities.validateCaps;
  }
});

require("source-map-support/register");

var _bluebird = _interopRequireDefault(require("bluebird"));

var _driver = require("./basedriver/driver");

var _core = require("./basedriver/core");

var _deviceSettings = require("./basedriver/device-settings");

var _protocol = require("./protocol");

Object.keys(_protocol).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _protocol[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _protocol[key];
    }
  });
});

var _constants = require("./constants");

var _static = require("./express/static");

var _server = require("./express/server");

var _proxy = require("./jsonwp-proxy/proxy");

var _status = require("./jsonwp-status/status");

var _capabilities = require("./basedriver/capabilities");

var _websocket = require("./express/websocket");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

_bluebird.default.config({
  cancellation: true
});

var _default = _driver.BaseDriver;
exports.default = _default;
