// @appium/docutils/build/lib/jsdoc-config-base.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("source-map-support/register");

var _path = _interopRequireDefault(require("path"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const docdashEntry = require.resolve('docdash');

const docdashPath = _path.default.dirname(docdashEntry);

const baseConfig = {
  plugins: ['plugins/markdown', 'jsdoc-plugin-typescript'],
  opts: {
    destination: './jsdoc-out',
    tutorials: './docs',
    source: {
      include: ['./lib'],
      exclude: [],
      includePattern: '.+\\.js(doc|x)?$',
      excludePattern: '(^|\\/|\\\\)_'
    },
    template: docdashPath,
    encoding: 'utf8'
  },
  typescript: {
    moduleRoot: 'lib'
  },
  docdash: {
    sectionOrder: ['Tutorials', 'Classes', 'Modules', 'Externals', 'Events', 'Namespaces', 'Mixins', 'Interfaces'],
    navLevel: 5,
    search: true,
    collapse: false
  }
};
var _default = baseConfig;
exports.default = _default;
