// @appium/base-driver/build/lib/basedriver/commands/find.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FindMixin = FindMixin;

require("source-map-support/register");

var _protocol = require("../../protocol");

function FindMixin(Base) {
  class FindCommands extends Base {
    async findElement(strategy, selector) {
      return await this.findElOrElsWithProcessing(strategy, selector, false);
    }

    async findElements(strategy, selector) {
      return await this.findElOrElsWithProcessing(strategy, selector, true);
    }

    async findElementFromElement(strategy, selector, elementId) {
      return await this.findElOrElsWithProcessing(strategy, selector, false, elementId);
    }

    async findElementsFromElement(strategy, selector, elementId) {
      return await this.findElOrElsWithProcessing(strategy, selector, true, elementId);
    }

    async findElOrEls(strategy, selector, mult, context) {
      throw new _protocol.errors.NotImplementedError('Not implemented yet for find.');
    }

    async getPageSource() {
      throw new _protocol.errors.NotImplementedError('Not implemented yet for find.');
    }

    async findElOrElsWithProcessing(strategy, selector, mult, context) {
      this.validateLocatorStrategy(strategy);

      try {
        return await this.findElOrEls(strategy, selector, mult, context);
      } catch (err) {
        if (this.opts.printPageSourceOnFindFailure) {
          const src = await this.getPageSource();
          this.log.debug(`Error finding element${mult ? 's' : ''}: ${err.message}`);
          this.log.debug(`Page source requested through 'printPageSourceOnFindFailure':`);
          this.log.debug(src);
        }

        throw err;
      }
    }

  }

  return FindCommands;
}
